scrapy/pipelines/media.py

```diff
@@ -112,12 +112,18 @@ def _process_request(self, request, info, item):
         info.downloading.add(fp)
         dfd = mustbe_deferred(self.media_to_download, request, info, item=item)
         dfd.addCallback(self._check_media_to_download, request, info, item=item)
+        dfd.addErrback(self._log_error, info)
         dfd.addBoth(self._cache_result_and_execute_waiters, fp, info)
-        dfd.addErrback(lambda f: logger.error(
-            f.value, exc_info=failure_to_exc_info(f), extra={'spider': info.spider})
-        )
         return dfd.addBoth(lambda _: wad)  # it must return wad at last
 
+    def _log_error(self, result, info):
+        logger.error(
+            result.value,
+            exc_info=failure_to_exc_info(result),
+            extra={'spider': info.spider}
+        )
+        return result
```

Comment:
Is this the only actual change?

Reply:
Well, the main change is actually moving the `addErrback` 1 line above. Returning `result` here is just so that the line move does not break the `addBoth` line that now goes afterwards.

Reply:
Oh, I see.

Reply:
The reason for the line move is that `_cache_result_and_execute_waiters` removes the failure information to save memory. So we were basically removing an error and then logging it, now we should be logging it and _then_ removing it.